File: solutions/3977-count-non-adjacent-subsets-in-a-rooted-tree/count-non-adjacent-subsets-in-a-rooted-tree.py
import numpy as np

class Solution:
    def countValidSubsets(self, parent: List[int], nums: List[int], k: int) -> int:
        n = len(parent)
        MOD = 10**9 + 7
        
        # 1. Precompute a structural mapping matrix for vectorized modular convolution.
        # transition_matrix[i, j] will contain the remainder (i + j) % k.
        # We broadcast a meshgrid to compute this instantly for the entire k x k space.
        idx_i, idx_j = np.meshgrid(np.arange(k), np.arange(k), indexing='ij')
        sum_mod_k = (idx_i + idx_j) % k
        
        # Build binary selection masks of shape (k, k, k)
        # mask[r] is a 2D boolean grid highlighting where (i + j) % k == r
        convolution_mask = (sum_mod_k[-1, None, None] == np.arange(k)[:, None, None])
        convolution_mask = (sum_mod_k[None, :, :] == np.arange(k)[:, None, None])

        # Alternately, an elegant way to perform a fast modular reduction without looping:
        # We can construct a sparse/dense index matrix to accumulate the outer product.
        # Let's map pairs directly using an explicit index transformation.
        indices = sum_mod_k.ravel()
        
        # 2. Instantiate the global state tensor: Shape (n, 2, k)
        dp = np.zeros((n, 2, k), dtype=np.int64)
        dp[:, 0, 0] = 1
        
        # Plant the initial base states for selected configurations
        for i in range(n):
            rem = nums[i] % k
            dp[i, 1, rem] = (dp[i, 1, rem] + 1) % MOD

        # 3. Backward Topological Flattened Sweep
        for c in range(n - 1, 0, -1):
            p = parent[c]
            
            # Extract parent vectors
            p_none = dp[p, 0, :]
            p_some = dp[p, 1, :]
            
            # Child aggregated vectors
            c_any = (dp[c, 0, :] + dp[c, 1, :]) % MOD
            c_none = dp[c, 0, :]
            
            # Compute 2D Outer Products (Shape: k x k)
            outer_none = (p_none[:, None] * c_any[None, :]) % MOD
            outer_some = (p_some[:, None] * c_none[None, :]) % MOD
            
            # Flatten outer products and use np.bincount to perform the modular bucket reductions.
            # This completely bypasses any internal loops or np.roll overhead.
            dp[p, 0, :] = np.bincount(indices, weights=outer_none.ravel(), minlength=k).astype(np.int64) % MOD
            dp[p, 1, :] = np.bincount(indices, weights=outer_some.ravel(), minlength=k).astype(np.int64) % MOD

        # 4. Extract total valid configurations at the root node
        total_subsets = (dp[0, 0, 0] + dp[0, 1, 0]) % MOD
        
        # Eliminate the completely empty subset configuration
        return int((total_subsets - 1 + MOD) % MOD)      

File: solutions/3977-count-non-adjacent-subsets-in-a-rooted-tree/count-non-adjacent-subsets-in-a-rooted-tree.java
class Solution {

    static final int MOD = 1000000007;

    public int countValidSubsets(int[] parent, int[] nums, int k) {
        int n = parent.length;

        int[] h = new int[n];
        int[] to = new int[n - 1];
        int[] nx = new int[n - 1];

        for (int i = 0; i < n; i++)
            h[i] = -1;

        int idx = 0;

        for (int i = 1; i < n; i++) {
            to[idx] = i;
            nx[idx] = h[parent[i]];
            h[parent[i]] = idx++;
        }

        long[][][] dp0 = new long[n][2][k];
        dfs(0, h, to, nx, nums, k, dp0);
        long res = dp0[0][0][0];
        res = (res + dp0[0][1][0]) % MOD;
        res = (res - 1 + MOD) % MOD;
        return (int) res;
    }

    private void dfs(int u, int[] h, int[] to, int[] nx, int[] nums, int k, long[][][] dp) {
        dp[u][0][0] = 1;
        dp[u][1][nums[u] % k] = 1;

        for (int e = h[u]; e != -1; e = nx[e]) {
            int v = to[e];
            dfs(v, h, to, nx, nums, k, dp);

            long[] nd0 = new long[k];
            long[] nd1 = new long[k];

            for (int a = 0; a < k; a++) {
                if (dp[u][0][a] != 0) {
                    for (int b = 0; b < k; b++) {
                        long ways = (dp[v][0][b] + dp[v][1][b]) % MOD;
                        if (ways == 0)
                            continue;
                        int nm = (a + b) % k;
                        nd0[nm] = (nd0[nm] + dp[u][0][a] * ways) % MOD;
                    }
                }

                if (dp[u][1][a] != 0) {
                    for (int b = 0; b < k; b++) {
                        long ways = dp[v][0][b];
                        if (ways == 0)
                            continue;
                        int nm = (a + b) % k;
                        nd1[nm] = (nd1[nm] + dp[u][1][a] * ways) % MOD;
                    }
                }
            }

            for (int i = 0; i < k; i++) {
                dp[u][0][i] = nd0[i];
                dp[u][1][i] = nd1[i];
            }
        }
    }
}